server.go

```diff
@@ -338,12 +341,65 @@ Loop:
 		}
 	}
 
-	s.log.Println("Waiting for in-flight connections...")
+	wg.Wait()
 	inflightWg.Wait()
-	s.log.Println("Bye")
 	return nil
 }
 
+// shutdown closes current clients and also stops accepting new clients in a
+// non-blocking manner
+func (s *Server) shutdown() {
+	s.manager.StopAcceptingConsumers()
```

Comment:
This is still racy (not trully blocking). As implemented the mutex protects the `teardown` variable, the `map` is not protected, so we still can end up with a new consumer when the `map` iteration takes place.

Reply:
Fixed.